azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,342 @@
+                              login=self.login, auth_type_dataplane=self.auth_type)
+
+    def upload_state(self, filename: str, replace: Optional[bool] = None):
+        '''
+        Uses device info from file to recreate the devices
+        '''
+
+        if replace:
+            self.delete_all_configs()
+            self.delete_all_devices()
+
+        hub_info = []
+        with open(filename, 'r', encoding='utf-8') as f:
+            for obj in f:
+                hub_info.append(json.loads(obj))
+
+        # upload configurations
+
+        configs = hub_info[0]
+
+        for i in tqdm(range(len(configs)), desc="Uploading hub configurations"):
+            c = configs[i]
+            iot_hub_configuration_create(cmd=self.cmd, config_id=c["id"], content=json.dumps(c["content"]), login=self.login,
+                                         target_condition=c["targetCondition"], priority=c["priority"], hub_name=self.hub_name,
+                                         metrics=json.dumps(c["metrics"]), resource_group_name=self.rg,
+                                         labels=json.dumps(c["labels"]), auth_type_dataplane=self.auth_type)
+
+        pbar = tqdm(total=len(hub_info), desc="Uploading devices and modules")
+        i = 1
+        pbar.update(1)
+
+        while(i < len(hub_info)):
+
+            # upload device identity and twin
+
+            numModules = hub_info[i]["numModules"]
+            identity = hub_info[i]
+            twin = hub_info[i + 1]
+
+            self.upload_device_identity(identity)
+
+            # all necessary twin attributes are already included in the identity. maybe don't save the twin and just do this:
+            # twin = identity
+            # if identity["authenticationType"] == DeviceAuthApiType.sas.value:
+            #     twin.pop("symmetricKey")
+
+            iot_device_twin_replace(cmd=self.cmd, device_id=identity["deviceId"], target_json=json.dumps(twin), login=self.login,
+                                    hub_name=self.hub_name, resource_group_name=self.rg, auth_type_dataplane=self.auth_type)
+
+            i += 2
+            pbar.update(2)
+
+            # upload module identities and twins for the given device
+
+            for _ in range(numModules):
+                module_identity = hub_info[i]
+                module_twin = hub_info[i + 1]
+
+                self.upload_module_identity(module_identity)
+
+                iot_device_module_twin_replace(cmd=self.cmd, device_id=identity["deviceId"], resource_group_name=self.rg,
+                                               module_id=module_identity["module_id"], target_json=json.dumps(module_twin),
+                                               hub_name=self.hub_name, login=self.login, auth_type_dataplane=self.auth_type)
+
+                i += 2
+                pbar.update(2)
+
+        pbar.close()
+
+        logger.info("Uploaded state from '{}' to IoT Hub '{}'".format(filename, self.hub_name))
+
+    def migrate_devices(
```

Comment:
if you rework how you export the state (specifically get everything you need before you write to the file), you can turn that into a helper function that would help you out here